### PR TITLE
Be more clear with errors found within anti-affinity rules.

### LINKED ISSUE #286: [BUG] no-anti-affinity rules does not work as expected
**System info:**
`stackrox/kube-linter:0.2.6`

**Describe the bug**
Even with anti-affinity rules apply to the pod, the `no-anti-affinity` rule fails

**To Reproduce**
```shell
# [Optional] Generate yaml manifest
helm repo add bitnami https://charts.bitnami.com/bitnami
helm template nginx bitnami/nginx --set replicaCount=2 --set containerSecurityContext.enabled=true --set containerSecurityContext.readOnlyRootFilesystem=true --set resources.requests.cpu=100m --set resources.limits.cpu=100m --set resources.requests.memory=100Mi  --set resources.limits.memory=100Mi > temp/temp.yaml

# Run kube linter
docker run --rm -v ${PWD}/temp:/data stackrox/kube-linter:0.2.6 lint /data/temp.yaml
```

**Sample YAML input**
```yaml
# Source: nginx/templates/deployment.yaml (stripped back)
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 2
  selector:
    matchLabels:
      app.kubernetes.io/name: nginx
      app.kubernetes.io/instance: nginx
  template:
    metadata:
      labels:
        app.kubernetes.io/name: nginx
        app.kubernetes.io/instance: nginx
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/name: nginx
                    app.kubernetes.io/instance: nginx
                namespaces:
                  - "default"
                topologyKey: kubernetes.io/hostname
              weight: 1

      containers:
        - name: nginx
          image: docker.io/bitnami/nginx:1.21.6-debian-10-r0
          securityContext:
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1001
          resources:
            limits:
              cpu: 100m
              memory: 100Mi
            requests:
              cpu: 100m
              memory: 100Mi

```

**Expected behavior**
All rules should pass

**Screenshots**
```shell
KubeLinter 0.2.6
/data/temp.yaml: (object: <no namespace>/nginx apps/v1, Kind=Deployment) object has 2 replicas but does not specify inter pod anti-affinity (check: no-anti-affinity, remediation: Specify anti-affinity in your pod specification to ensure that the orchestrator attempts to schedule replicas on different nodes. Using podAntiAffinity, specify a labelSelector that matches pods for the deployment, and set the topologyKey to kubernetes.io/hostname. Refer to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity for details.)
```

**Additional context**
When `replicas` is set to 1, the rule correctly passes

### PR DESCRIPTION
**Pull request by dhaus67**
Fixes #286 .

Now, more specific errors will be returned when running checks against anti-affinity rules that have misconfigurations:

- Pod's and pod affinity's namespaces are not matching: `pod's namespace "pod-namespace" not found in anti-affinity's namespaces [affinity-namespace]`
- Topology key does not match: `anti-affinity's topology key does not match "some.key/value"`
- Pod's and pod affinity's labels are not matching: `pod's labels \"pod=labels\" do not match with anti-affinity's labels \"anti-affinity=labels\"`

### CODE REVIEW

File: pkg/templates/antiaffinity/template.go [hunk 1]
```diff
@@ -28,9 +29,10 @@ func defaultTopologyKeyMatcher(topologyKey string) bool {
 
 func init() {
 	templates.Register(check.Template{
-		HumanName:   "Anti affinity not specified",
-		Key:         templateKey,
-		Description: "Flag objects with multiple replicas but inter-pod anti affinity not specified in the pod template spec",
+		HumanName: "Anti affinity not specified",
+		Key:       templateKey,
+		Description: "Flag objects with multiple replicas but inter-pod anti affinity not specified in the pod " +
+			"template spec",
```
Comment (janisz) [inline pkg/templates/antiaffinity/template.go:35]: Why this change?

File: pkg/templates/antiaffinity/template.go [hunk 2]
```diff
@@ -59,30 +61,47 @@ func init() {
...
 						}
+						foundIssues = append(foundIssues, diagnostic.Diagnostic{
+							Message: err.Error(),
+						})
 					}
 					for _, required := range requiredAffinity {
-						if affinityTermMatchesLabelsAgainstNodes(required, podTemplateSpec.Namespace, podTemplateSpec.Labels, topologyKeyMatcher) {
+						err := affinityTermMatchesLabelsAgainstNodes(required, podTemplateSpec.Namespace,
+							podTemplateSpec.Labels, topologyKeyMatcher)
+						if err == nil {
```
Comment (janisz) [inline pkg/templates/antiaffinity/template.go:81]: Should we do a short circuit here? What if we have an error in next iteration?

File: pkg/templates/antiaffinity/template.go [hunk 3]
```diff
@@ -59,30 +61,47 @@ func init() {
...
 							return nil
 						}
+						foundIssues = append(foundIssues, diagnostic.Diagnostic{
+							Message: err.Error(),
+						})
 					}
 				}
+				if foundIssues != nil {
```
Comment (janisz) [inline pkg/templates/antiaffinity/template.go:89]: I think this should be in above `if` body

File: pkg/templates/antiaffinity/template.go [hunk 4]
```diff
@@ -59,30 +61,47 @@ func init() {
 				if !hasPods {
 					return nil
 				}
+				var foundIssues []diagnostic.Diagnostic
 				if affinity := podTemplateSpec.Spec.Affinity; affinity != nil && affinity.PodAntiAffinity != nil {
```
Comment (janisz) [inline pkg/templates/antiaffinity/template.go:65]: How about inverting this condition?

**Review by janisz — CHANGES_REQUESTED** [review]
I left some comments

File: pkg/templates/antiaffinity/template.go (hunk 1 above)
Comment (dhaus67) [reply pkg/templates/antiaffinity/template.go:35]: Refactored to not exceed line-length of 130 characters.

File: pkg/templates/antiaffinity/template.go (hunk 2 above)
Comment (dhaus67) [reply pkg/templates/antiaffinity/template.go:81]: I was wondering the same thing, I've kept the logic as before, short-circuiting on the first found, valid anti-affinity rule.
I'm fine with changing it to _not_ short-circuit, if we are fine with it.

File: pkg/templates/antiaffinity/template.go (hunk 3 above)
Comment (janisz) [reply pkg/templates/antiaffinity/template.go:93]: Now this condition is redundant

File: pkg/templates/antiaffinity/template.go [hunk 5]
```diff
@@ -59,30 +61,48 @@ func init() {
...
 				}
+				return nil
 			}, nil
 		}),
 	})
 }
 
-func affinityTermMatchesLabelsAgainstNodes(affinityTerm coreV1.PodAffinityTerm, podNamespace string, podLabels map[string]string, topologyKeyMatcher func(string) bool) bool {
-	// If namespaces is not specified in the affinity term, that means the affinity term implicitly applies to the pod's namespace.
+func affinityTermMatchesLabelsAgainstNodes(affinityTerm coreV1.PodAffinityTerm, podNamespace string,
```
Comment (viswajithiii) [inline pkg/templates/antiaffinity/template.go:102]: Nit: this function should be renamed now that it's no longer returning a `bool`. Perhaps `validateAffinityTermMatchesAgainstNodes`?

**Review by viswajithiii — APPROVED** [review]

File: pkg/templates/antiaffinity/template.go (hunk 3 above)
Comment (dhaus67) [reply pkg/templates/antiaffinity/template.go:93]: Screenshot 2022-05-05 at 18 06 23

**Pull request merged**